### DISCUSSION TITLE
Difference from FastUI and Template Engines

### DISCUSSION THREAD

**Discussion by hasansezertasan** [paveldedik/ludic#34] [General]
The purpose of this question is mainly motivational.

## Thoughts about FastUI

FastUI is growing and is an interesting package, it claims to be "A new way to build web application UIs" but there are quite a few similar attempts (like [amis](https://github.com/baidu/amis) and [layui](https://github.com/layui/layui)).

I've been playing with it since it came out and it was very helpful to avoid JS, HTML, and CSS and stay in a Python environment. 

There is a but, I never liked the "React" bridge. I was not able to interact with the DOM via Vanilla JS (in any abstraction level, you always need to inject some sort of script or override something for a darn reason...)

## Thoughts about Template Engines

I'm bringing Template Engines question since Ludic looks like a direct alternative to many Template Engines in the web development scope.

I LOVE PYTHON. I want to do my business in a `.py` 🐍 file, not an HTML dot JINJA dot IDK_WHAT_IS_NEXT file.

Template engines bring their syntax, and they mostly lack editor support and auto-completion. Understanding their structure (thanks to awkward abstractions) is time-consuming.

Template Engines come with a nightmare; "micro-syntaxes" 😮‍💨.

## Thoughts about Ludic

I also do love web dev but have never enjoyed so-called "Modern Web Dev" (might be because of JSX micro-syntax and bundling).

HTMX was and is a life savior, but I'm still writing HTML...

This package, Ludic, is the thing I've been dreaming (no I'm not a cyborg 🤖) of for two years now (check the motivation section at [awesome-python-html](https://github.com/hasansezertasan/awesome-python-html/?tab=readme-ov-file#motivation)) Thanks to everyone who worked on this project 🙏.

## Comparing with FastUI

* In both FastUI and Ludic, you are not leaving sweet sweet Python.
* FastUI comes with a React App Bundle. It brings ReactJS abstraction to the DOM. ReactJS knowledge is required to override some of its features. It's not easy to get your hands dirty. For a new HTML tag, a component, or an integration with a third-party JavaScript library, you'll probably be dependent on FastUI developers.
* Ludic is generating HTML strings, with no abstraction over the DOM. It gives you the ability to interact with the DOM via HTMX, Vanilla JS, or any basic web dev tool. It's easy to get your hands dirty.
* FastUI is a CSR tech.
* Ludic is an SSR tech.
* FastUI Depends on FastAPI.
* Ludic doesn't depend on any framework.
* FastUI is like a sandbox inside another (browser).
* Ludic is another (pythonic) way to template HTML fragments.

## Comparing with Template Engines

* Template Engines bring micro-syntaxes.
* Ludic is all Python 🐍.
* Template Engines are hard to override.
* Ludic is all Python 🐍.
* Template Engines are limiting...
* Ludic is all Python 🐍.

## Summary

I believe Ludic is a promising piece of tech and deserves a more in-depth comparison against many other tools.

These are the points that caught my eye.

I'd like to hear more from the author, and other contributors.

**Comment by paveldedik** [comment]
Hi @hasansezertasan 

thank you very much for the kind words :-) This really motivates me to continue working on this project.

And you have a really good point about deserving a more in-depth comparison with other tools. I didn't really think about that until now, but that would also help people understand what Ludic does and how it differs from other tools. I'll definitely add that to my roadmap.

As for the contributors, it's only me for now. I was thinking about similar framework for a while and after I heard about HTMX, I thought it is time to give it a try. When I started working on it, I got really excited about how nicely it works and here we are - I have a lot stuff on my roadmap.

## What do I have on the Roadmap?

I'm currently mostly working on layout components inspired by https://every-layout.dev/ which seems like a huge productivity boost when creating websites. Creating components in ludic copying this layout approach, I am able to create responsive HTML pages literally in a couple of minutes.

I'm also creating a more interactive documentation so that people can quickly test how Ludic works. It is too limiting to use only readthedocs.

And I also want to create a cookiecutter so that getting started with the framework is as fast as possible.

The most potentially problematic thing I'm facing is https://github.com/paveldedik/ludic/issues/3
The problem is that attributes in components are not typed due to that - so even IDE suggestions don't work, which would be a really great feature (e.g. it is really nice how CSS properties are suggested by IDE, it would be even more powerful to have that for component attributes as well). The solution I'm considering is to create at least a mypy plugin until the issues is resolved by a PEP and adopted by type checkers.

And than there are more features from the HTMX library that should be tested, implemented in examples, and so on.

Than the framework requires more propagation, it would be nice to create a discourse, social networks and so on.

All in all, quite a lot of work, unfortunately I also have a full time job so it won't be that fast as I would like :)

**Comment by hasansezertasan** [comment]
> thank you very much for the kind words :-) This really motivates me to continue working on this project.

Happy to hear that.

> And you have a really good point about deserving a more in-depth comparison with other tools. I didn't really think about that until now, but that would also help people understand what Ludic does and how it differs from other tools. I'll definitely add that to my roadmap.

I think it will help a lot. I'm highly excited about the idea of building web interfaces using only Python but all the other options like GradIO, Reflex, NiceGUI, FastUI, etc. do depend on some sort of frontend framework like VueJS, ReactJS, Svelte, etc. while separating the connection between frontend development and backend development.

> As for the contributors, it's only me for now. I was thinking about similar framework for a while and after I heard about HTMX, I thought it is time to give it a try. When I started working on it, I got really excited about how nicely it works and here we are - I have a lot stuff on my roadmap.

I just noticed that. You've done a great job! Why don't you share your entire roadmap with us? There might be something an interested person could help.

> I'm currently mostly working on layout components inspired by https://every-layout.dev/ which seems like a huge productivity boost when creating websites. Creating components in ludic copying this layout approach, I am able to create responsive HTML pages literally in a couple of minutes.

The site you referred to looks quite interesting. Can you give an example of the process you just mentioned? Maybe a video or a small written tutorial?

> I'm also creating a more interactive documentation so that people can quickly test how Ludic works. It is too limiting to use only readthedocs.

I recommend trying MkDocs. I didn't try Ludic with PyScript but you can achieve something really interactive using pyodide. Checkout Pydantic documentation, it has really great presentation for examples.

> And I also want to create a cookiecutter so that getting started with the framework is as fast as possible.

Yet another recommendation is to use copier...

> The most potentially problematic thing I'm facing is https://github.com/paveldedik/ludic/issues/3
The problem is that attributes in components are not typed due to that - so even IDE suggestions don't work, which would be a really great feature (e.g. it is really nice how CSS properties are suggested by IDE, it would be even more powerful to have that for component attributes as well). The solution I'm considering is to create at least a mypy plugin until the issues is resolved by a PEP and adopted by type checkers.

Good luck with that!

> Than the framework requires more propagation, it would be nice to create a discourse, social networks and so on.

Let me know if you create one, especially a Discord channel could be really handy!

> All in all, quite a lot of work, unfortunately I also have a full time job so it won't be that fast as I would like :)

I totally get you 🫂. I also work full-time and check out my public repositories... Les then 20 percent of them are unfinished...

**Reply by paveldedik** [reply]
Thank you for answer and sorry it took so long for me to reply. 

> I just noticed that. You've done a great job! Why don't you share your entire roadmap with us? There might be something an interested person could help.

Thank you. Yeah I'll create one soon hopefully.

> The site you referred to looks quite interesting. Can you give an example of the process you just mentioned? Maybe a video or a small written tutorial?

Yeah I'm working on exacly that :) I want to make sure people understand the benefits and the potential. I first need to finish all the layouts.

> I recommend trying MkDocs. I didn't try Ludic with PyScript but you can achieve something really interactive using pyodide. Checkout Pydantic documentation, it has really great presentation for examples.

Yeah I'm using mkdocs right now but it doesn't allow me to do the kind of stuff I want. I want to create very interactive examples in future.

> Yet another recommendation is to use copier...

Thank you for the suggestion, I didn't know about this tool, I'll check it :)

> Let me know if you create one, especially a Discord channel could be really handy!

I'll make sure to lat you know :)

> I totally get you 🫂. I also work full-time and check out my public repositories... Les then 20 percent of them are unfinished...

Yeah after already working 40 hours a week, it is sometimes hard to have motivation for even more work :)

**Reply by hasansezertasan** [reply]
Thank you for your answer. It should be me who apologizes for being so late about latency.

> Yeah I'm working on exacly that :) I want to make sure people understand the benefits and the potential. I first need to finish all the layouts.

Looking forward to see it. 🤓 

> Yeah I'm using mkdocs right now but it doesn't allow me to do the kind of stuff I want. I want to create very interactive examples in future.

I see, don't forget to start a discussion on it, I'll get notified because I'm watching the repository. 📣

> Yeah after already working 40 hours a week, it is sometimes hard to have motivation for even more work :)

Well, over-working sometimes feels like therapy... 😆.

**Comment by paveldedik** [comment]
Hi @hasansezertasan 

What do you think about a comparison table like this? Does it make sense to you or would you remove/add something? Thank you again for the idea.

## Comparison

Here is a table comparing Ludic to other similar tools:

| Feature                    | Ludic       | FastUI   | Reflex      |
|----------------------------|-------------|----------|-------------|
| HTML rendering             | Server Side | Client Side | Client Side |
| Uses a template engine     | No          | No       | No          |
| UI interactivity           | [</> htmx](https://htmx.org)* | [React](https://react.dev/) | [React](https://react.dev/) |
| Backend framework          | [Starlette](https://www.starlette.io)*  | [FastAPI](https://fastapi.tiangolo.com) | [FastAPI](https://fastapi.tiangolo.com) |
| Client-Server Communication | [HTML + REST](https://htmx.org/essays/how-did-rest-come-to-mean-the-opposite-of-rest/) | [JSON + REST](https://github.com/pydantic/FastUI?tab=readme-ov-file#the-principle-long-version) | [WebSockets](https://reflex.dev/blog/2024-03-21-reflex-architecture/) |

<sup>(*) HTMX as well as Starlette are optional dependencies for Ludic, it does not enforce any frontend or backend frameworks. At it's core, Ludic only generates HTML and allows registering CSS.</sup>

**Reply by paveldedik** [reply]
Hi @hasansezertasan 

here is the discourse link if you are interested to join: https://discord.com/invite/4Y5fSQUS

**Reply by hasansezertasan** [reply]
Hey @paveldedik.

> What do you think about a comparison table like this? Does it make sense to you or would you remove/add something? Thank you again for the idea.

I think you have got everything right. Looking informative. I'd like to extend it in the future. I really like comparison charts, checkout jowilf/starlette-admin#333

> here is the discourse link if you are interested to join: https://discord.com/invite/4Y5fSQUS

Oh boi, you did it 🥳! I'm coming to the party 🎉!

**Reply by hasansezertasan** [reply]
I just saw the updates on the documentation, looking awesome! Good job ㊗️. I'll check it out ASAP.
